contracts/Curation.sol

```diff
@@ -0,0 +1,415 @@
+            staking.unlock(application.applicant, application.lockId);
+            staking.unlock(msg.sender, lockId);
+            return 0;
+        }
+
+        // create vote
+        // TODO: metadata
+        // script to call `resolveChallenge(entryId)`
+        uint256 scriptLength = 64; // 4 (spec) + 20 (address) + 4 (calldataLength) + 4 (signature) + 32 (input)
+        bytes4 spec = bytes4(0x01);
+        address target = address(this);
+        bytes memory targetBytes = new bytes(32);
+        bytes4 calldataLength = bytes4(0x24); // 4 + 32
+        bytes4 signature = this.resolveChallenge.selector;
+        bytes memory executionScript = new bytes(scriptLength);
+        // concatenate spec + address(this) + calldataLength + calldata
+        // TODO: should we put this somewhere in aragonOS to be able ti reuse? (if it's not already there)
+        assembly {
+            mstore(add(executionScript, 0x20), spec)
+            mstore(add(targetBytes, 0x20), target)
+            mstore(add(executionScript, 0x24), mload(add(targetBytes, 0x2C)))
+            mstore(add(executionScript, 0x38), calldataLength)
+            mstore(add(executionScript, 0x3C), signature)
+            mstore(add(executionScript, 0x40), entryId)
+        }
+        uint256 voteId = voting.newVote(executionScript, "");
+
+        challenges[entryId] = Challenge({
+            challenger: msg.sender,
+            date: getTimestamp(),
+            resolved: false,
+            amount: amount,
+            lockId: lockId,
+            voteId: voteId,
+            dispensationPct: dispensationPct
+        });
+
+        NewChallenge(entryId, msg.sender);
+
+        return entryId;
+    }
+
+    function resolveChallenge(bytes32 entryId) isInitialized public {
+        Challenge storage challenge = challenges[entryId];
+        Application storage application = applications[entryId];
+
+        require(!challenge.resolved);
+        // to avoid resolving and redistributing twice
+        challenge.resolved = true;
+        // TODO: canExecute??
+        require(voting.isClosed(challenge.voteId));
+
+        bool voteResult;
+        (voteResult,,) = voting.getVoteResult(challenge.voteId);
+
+        uint256 reward;
+        if (voteResult == false) { // challenge not accepted (application remains)
+            // it's still in application phase (not registered yet)
+            if (!application.registered) {
+                // insert in Registry app
+                registry.add(application.data);
+                application.registered = true;
+            }
+            // Remove applicant (winner) used lock
+            delete(usedLocks[application.applicant][application.lockId]);
+            // Unlock challenger tokens from Staking app
+            reward = challenge.amount.mul(dispensationPct) / PCT_BASE;
+            // Redistribute tokens
+            staking.unlockAndMoveTokens(challenge.challenger, challenge.lockId, application.applicant, reward);
```

Comment:
Also an issue with staking, but it would be more descriptive if the function signature referenced the fact that the unlock is partial: `staking.partialUnlockAndMove(...)`